src/gluonts/model/n_beats/_network.py

```diff
@@ -0,0 +1,521 @@
+        )
+        flag = seasonal_error == 0
+
+        mase = (
+            F.mean(F.abs(future_target - forecast), axis=1) * (1 - flag)
+        ) / (seasonal_error + flag)
+
+        return mase
+
+
+class NBEATSTrainingNetwork(NBEATSNetwork):
+    @validated()
+    def __init__(self, loss_function: str, freq: str, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.loss_function = loss_function
+        self.freq = freq
+
+        # Covert frequency string like "2H" to whats called the periodicity m.
+        # E.g. 12 in case of "2H" because of 12 times two hours in a day.
+        self.periodicity = get_seasonality(self.freq)
+
+    # noinspection PyMethodOverriding,PyPep8Naming
+    def hybrid_forward(
+        self, F, past_target: Tensor, future_target: Tensor
+    ) -> Tensor:
+        """
+
+        Parameters
+        ----------
+        F
+        past_target
+            Tensor with past observations.
+            Shape: (batch_size, context_length, target_dim).
+        future_target
+            Tensor with future observations.
+            Shape: (batch_size, prediction_length, target_dim).
+
+        Returns
+        -------
+        Tensor
+            Loss tensor. Shape: (batch_size, ).
+        """
+        # future_target never used
+        forecast = super().hybrid_forward(
+            F, past_target=past_target, future_target=future_target
+        )
+
+        if self.loss_function == "sMAPE":
+            loss = self.smape_loss(F, forecast, future_target)
+        elif self.loss_function == "MAPE":
+            loss = self.mape_loss(F, forecast, future_target)
+        elif self.loss_function == "MASE":
+            loss = self.mase_loss(
+                F, forecast, future_target, past_target, self.periodicity
+            )
+        else:
+            raise ValueError(
+                f"Invalid value {self.loss_function} for argument loss_function."
+            )
+
+        return loss
+
+
+class NBEATSPredictionNetwork(NBEATSNetwork):
+    @validated()
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+    # noinspection PyMethodOverriding,PyPep8Naming
+    def hybrid_forward(
+        self, F, past_target: Tensor, future_target: Tensor = None
+    ) -> Tensor:
+        """
+
+        Parameters
+        ----------
+        F
+        past_target
+            Tensor with past observations.
+            Shape: (batch_size, context_length, target_dim).
+        future_target
+            Not used.
+
+        Returns
+        -------
+        Tensor
+            Prediction sample. Shape: (samples, batch_size, prediction_length).
+        """
+        # future_target never used
+        forecasts = super().hybrid_forward(
+            F, past_target=past_target, future_target=past_target
+        )
+        # dimension collapsed previously because we only have one sample each:
+        forecasts = F.expand_dims(forecasts, axis=1)
```

Comment:
Should it be `axis=0` here? I mean in order to get the shape as in the docstring (which I believe is the correct one that the `SampleForecast` class expects)

Reply:
No, it has to be 1 here. Axis one is the num_samples axis. The axis is removed somewhere if there is only one sample, this leads to dimensionality errors down the line.